store/store.go

```diff
@@ -92,50 +96,41 @@ func (bs *BlockStore) LoadBaseMeta() *types.BlockMeta {
 	return bs.LoadBlockMeta(bs.base)
 }
 
-// LoadBlock returns the block with the given height.
-// If no block is found for that height, it returns nil.
-func (bs *BlockStore) LoadBlock(height int64) *types.Block {
-	var blockMeta = bs.LoadBlockMeta(height)
+// LoadBlock fetches the block at the provided height from IPFS and the local db
+func (bs *BlockStore) LoadBlock(ctx context.Context, height int64) (*types.Block, error) {
+	blockMeta := bs.LoadBlockMeta(height)
 	if blockMeta == nil {
-		return nil
+		// TODO(evan): return an error
+		return nil, nil
 	}
 
-	pbb := new(tmproto.Block)
-	buf := []byte{}
-	for i := 0; i < int(blockMeta.BlockID.PartSetHeader.Total); i++ {
-		part := bs.LoadBlockPart(height, i)
-		// If the part is missing (e.g. since it has been deleted after we
-		// loaded the block meta) we consider the whole block to be missing.
-		if part == nil {
-			return nil
-		}
-		buf = append(buf, part.Bytes...)
-	}
-	err := proto.Unmarshal(buf, pbb)
+	lastCommit := bs.LoadBlockCommit(height - 1)
+
+	data, err := ipld.RetrieveBlockData(ctx, &blockMeta.DAHeader, bs.dag, rsmt2d.NewRSGF8Codec())
 	if err != nil {
-		// NOTE: The existence of meta should imply the existence of the
-		// block. So, make sure meta is only saved after blocks are saved.
-		panic(fmt.Sprintf("Error reading block: %v", err))
+		return nil, err
```

Comment:
We need to have an error check here to return a meaningful error to the caller in case a block is not found __locally__

Reply:
Just for my understanding: if the dag passed in dag has the offline mode enabled, is that sufficient to ensure that we failed to find the block *locally*? Or is there some sentinel error or error type to check against?

Reply:
good point, resolved here https://github.com/lazyledger/lazyledger-core/pull/374/commits/507f04d38f3358c6aebfa89568db10c4f24f3241

Reply:
> Just for my understanding: if the dag passed in dag has the offline mode enabled, is that sufficient to ensure that we failed to find the block locally? Or is there some sentinel error or error type to check against?

That's a great point and I think we should have a test that ensures that the correct error is thrown and the timeout is not reached. Added here https://github.com/lazyledger/lazyledger-core/pull/374/commits/51c7f44b30ae96d439b047292bd7355dc5eb3476

Reply:
> if the dag passed in dag has the offline mode enabled, is that sufficient to ensure that we failed to find the block locally? 
I guess so, but there could be other errors as well, e.g. related to erasure coding. Also, retrieval code has its own error type, but usually, for DAG-related not found errors I check for ipld.ErrNotFound.